[lib/Sema] Suggest `return` when the last statement would be a valid …

moritzdietsche opened the pull request:
…return value

<!-- What's in this pull request? -->
When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

Note: SR-10988 proposes to extend this behavior (for non-opaque and opaque return types) further by looking for valid return values in `if` statements. I do not fully understand the benefits of having this but will gladly provide an implementation if it is decided that such functionality is desired.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves #53378

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

LucianoPAlmeida commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2645,6 +2645,46 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
     // If there are no candidates, then the body has no return statements, and
     // we have nothing to infer the underlying type from.
     if (Candidates.empty()) {
+      // We try to find valid candidates for return values to see if we can offer
+      // a fixit before presenting the default diagnosis.
```
```suggestion
      // We try to find if the last element of the `Body` multi element `BraceStmt` is an expression that produces a value that satisfies all the opaque type requirements and if that is the case, it means we can suggest a fix-it note to add an explicit `return`. 
```

LucianoPAlmeida commented on test/type/opaque.swift:
```diff
@@ -533,3 +533,15 @@ func test_diagnostic_with_contextual_generic_params() {
+    func test() -> some Numeric {
+      // expected-error@-1 {{function declares an opaque return type, but has no return statements in its body from which to infer an underlying type}}
+      let x = 0
+      x // expected-note {{did you mean to return the last expression?}}
```
We should make sure the correct fix-it is produced as well
```suggestion
      x // expected-note {{did you mean to return the last expression?}} {{7-7=return }}
```

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

moritzdietsche commented:
Thank you @LucianoPAlmeida for your review. I implemented the suggested changes in the latest commit.

LucianoPAlmeida commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2645,6 +2645,46 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
+          for (auto requirement :
+               OpaqueDecl->getOpaqueInterfaceGenericSignature().getRequirements()) {
+            auto requirementFulfilled =
+                TypeChecker::typesSatisfyConstraint(expr->getType()->getRValueType(),
```
Ah, there is a `TypeChecker::conformsToProtocol` maybe could be simpler to use here?

moritzdietsche replied:
It saved one line, must have overlooked this one. Thanks again!

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

xedin commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2645,6 +2645,44 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
+          bool conforms = true;
+
+          for (auto requirement :
+               OpaqueDecl->getOpaqueInterfaceGenericSignature().getRequirements()) {
```
We cannot assume that all of the requirements are going to be protocol requirements. Let's switch to use `llvm::all_of` and check that requirement is `RequirementKind::Conformance` before calling `getProtocolDecl()` on it because it would crash the compiler otherwise.

moritzdietsche replied:
Thanks for catching this. I have chosen not to attempt to continue if there are any other requirements. Please advise if this is overly defensive.

LucianoPAlmeida commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2645,6 +2645,44 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
+
+      // We have not found any candidates for where adding a `return` would fix
+      // the issue and are therefore falling back to the generic error message.
       Implementation->diagnose(diag::opaque_type_no_underlying_type_candidates);
```
Also, maybe we could move this logic here... since diagnostic will always be emitted we can put this logic after it and if conditions are met we emit only the fix-it :)

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

moritzdietsche committed:
Merge branch 'apple:main' into SR-10988-Suggest-return-when-the-last-statement-of-a-multi-statement-function-body-would-be-a-valid-return-value

xedin reviewed (commented):
I think the check makes sense. Could you please add some more tests where e.g. `some` type is a protocol composition  and one case where result doesn't conform?

moritzdietsche committed:
Merge branch 'apple:main' into SR-10988-Suggest-return-when-the-last-statement-of-a-multi-statement-function-body-would-be-a-valid-return-value

moritzdietsche committed:
Merge branch 'apple:main' into SR-10988-Suggest-return-when-the-last-statement-of-a-multi-statement-function-body-would-be-a-valid-return-value

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

xedin commented on test/type/opaque.swift:
```diff
@@ -543,5 +547,23 @@ func sr10988() {
+  struct S3: P1 {
+    func test3() -> some P1 & P2 {
+      // expected-error@-1 {{function declares an opaque return type, but has no return statements in its body from which to infer an underlying type}}
+      let x = S2() // expected-warning {{initialization of immutable value 'x' was never used; consider replacing with assignment to '_' or removing it}}
```
I think you'd want to have `x` as a last expression here as well so the logic you added is exercised.

moritzdietsche replied:
Right, will just fix it too.

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

xedin commented:
@swift-ci please test

moritzdietsche commented:
> Looks great now, thank you @moritzdietsche!

Thanks for your review and help!

LucianoPAlmeida commented:
@swift-ci Please test Linux Platform

xedin commented:
@moritzdietsche Looks like you need to update one more test.

LucianoPAlmeida commented:
Ah, sorry @xedin ... I just looked the Jenkins and saw `No identified problem` so I thought it was some flaky failure and re-kicked it. Should've looked at the logs regardless

xedin commented:
No worries! :)

moritzdietsche committed:
[lib/Sema] Suggest `return` when the last statement would be a valid return value

Addresses SR-10988. When a multi-statement function body contains no return keyword but the last statement would be a valid return value, the compiler currently proposes a fixit to add the return keyword.
This changes extends this behavior to functions with opaque return types.

moritzdietsche commented:
@LucianoPAlmeida I think I'm ready for a new test run.

LucianoPAlmeida commented:
@swift-ci Please test
